clippy_lints/src/functions/mod.rs

```diff
@@ -399,6 +400,50 @@ declare_clippy_lint! {
     "renamed function parameters in trait implementation"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Warns when a function signature uses `&Option<T>` instead of `Option<&T>`.
+    ///
+    /// ### Why is this bad?
+    /// More flexibility, better memory optimization, and more idiomatic Rust code.
+    ///
+    /// `&Option<T>` in a function signature breaks encapsulation because the caller must own T
+    /// and move it into an Option to call with it. When returned, the owner must internally store
+    /// it as `Option<T>` in order to return it.
+    /// At a lower level `&Option<T>` points to memory that has `presence` bit flag + value,
+    /// whereas `Option<&T>` is always optimized to a single pointer.
```

Comment:
Question: is this *always* guaranteed, or is it only guaranteed if `T: Sized` (I haven't looked very hard myself)?

Reply:
(I haven't looked into this too far; please take everything I say with a grain of salt)

[`Option`'s documentation in `std`](https://doc.rust-lang.org/1.81.0/std/option/index.html#representation) only guarantees this for `&{,mut} impl Sized`, but it seems like [the Unsafe Code Guidelines Working Group tenatively posit that similar optimizations occur for all values with a niche](https://rust-lang.github.io/unsafe-code-guidelines/layout/enums.html#discriminant-elision-on-option-like-enums), which would include `&{,mut} impl !Sized` (as [all references must be nonnull, even if their pointee is `!Sized`](https://doc.rust-lang.org/1.81.0/std/primitive.reference.html#safety)).

Obviously, though, `Option<&impl`[`!Thin`](https://doc.rust-lang.org/1.81.0/core/ptr/traitalias.Thin.html)`>` cannot be optimized to a thin pointer, and there does not seem to be any guarantee about what the metadata component would be in `None::<&impl !Thin>`.

Reply:
Yeah, which is why I asked, because I wasn't sure if this is always the case.

Reply:
The real question is if we want to have all this complexity explained in a lint? I agree we should link to the options doc, and perhaps talk about **possible** performance improvement.  Any suggestions on the wording?

An even more important question is:  how common is a case where API is deliberately, as oppose to accidentally, designed to take or return `&Option<T>`? If this is extremely rare, I think it makes sense to promote this to `style`, at least in the longer term.